// @antv/l7plot/dist/lib/plots/grid/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Grid = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const grid_layer_1 = require("../../layers/grid-layer");
const layer_group_1 = require("../../core/layer/layer-group");
class Grid extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = plot_1.Plot.PlotType.Grid;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Grid.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.gridLayer = new grid_layer_1.GridLayer(Object.assign({ source }, (0, util_1.pick)(this.options, grid_layer_1.GridLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.gridLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(this.source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const heatMapLayerConfig = (0, util_1.pick)(options, grid_layer_1.GridLayer.LayerOptionsKeys);
        this.gridLayer.update(heatMapLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
}
exports.Grid = Grid;
/**
 * 默认配置项
 */
Grid.DefaultOptions = constants_1.DEFAULT_OPTIONS;
