// @antv/l7plot/dist/lib/plots/heatmap/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Heatmap = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const heatmap_layer_1 = require("../../layers/heatmap-layer");
const layer_group_1 = require("../../core/layer/layer-group");
class Heatmap extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = plot_1.Plot.PlotType.Heatmap;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Heatmap.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.heatmapLayer = new heatmap_layer_1.HeatmapLayer(Object.assign({ source }, (0, util_1.pick)(this.options, heatmap_layer_1.HeatmapLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.heatmapLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(this.source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const heatMapLayerConfig = (0, util_1.pick)(options, heatmap_layer_1.HeatmapLayer.LayerOptionsKeys);
        this.heatmapLayer.update(heatMapLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
    /**
     * 实现 legend 配置项
     */
    getLegendOptions() {
        var _a;
        const sizeLegendItems = this.heatmapLayer.layer.getLegendItems('size');
        if (Array.isArray(sizeLegendItems) && sizeLegendItems.length !== 0) {
            const min = sizeLegendItems[0].value;
            const max = sizeLegendItems[sizeLegendItems.length - 1].value;
            const colors = (_a = this.heatmapLayer.options.style) === null || _a === void 0 ? void 0 : _a.colorsRamp.map(({ color }) => color);
            return { type: 'continue', min, max, colors };
        }
        return {};
    }
}
exports.Heatmap = Heatmap;
/**
 * 默认配置项
 */
Heatmap.DefaultOptions = constants_1.DEFAULT_OPTIONS;
