// @antv/l7plot/dist/lib/layers/prism-layer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismLayer = void 0;
const util_1 = require("@antv/util");
const l7_1 = require("@antv/l7");
const plot_layer_1 = require("../../core/layer/plot-layer");
const adaptor_1 = require("./adaptor");
const helper_1 = require("../dot-layer/helper");
const DEFAULT_OPTIONS = {};
const LAYER_OPTIONS_KEYS = ['color', 'size', 'style', 'state'];
class PrismLayer extends plot_layer_1.PlotLayer {
    constructor(options) {
        super(options);
        /**
         * 图层类型
         */
        this.type = plot_layer_1.PlotLayer.LayerType.PrismLayer;
        /**
         * 图层是否具有交互属性
         */
        this.interaction = true;
        const { name, source } = this.options;
        const config = this.pickLayerConfig(this.options);
        this.name = name ? name : (0, util_1.uniqueId)(this.type);
        this.layer = new l7_1.PolygonLayer(Object.assign(Object.assign({}, config), { name: this.name }));
        this.mappingLayer(this.layer, this.options);
        this.setSource(source);
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return DEFAULT_OPTIONS;
    }
    mappingLayer(layer, options) {
        (0, adaptor_1.mappingLayer)(layer, options);
    }
    update(options) {
        super.update(options);
        this.mappingLayer(this.layer, this.options);
    }
    getColorLegendItems() {
        const colorLegendItems = this.layer.getLegendItems('color');
        if (Array.isArray(colorLegendItems) && colorLegendItems.length !== 0) {
            const items = (0, helper_1.getColorLegendItems)(colorLegendItems);
            return items;
        }
        return [];
    }
}
exports.PrismLayer = PrismLayer;
/**
 * 默认配置项
 */
PrismLayer.DefaultOptions = DEFAULT_OPTIONS;
/**
 * 图层配置项 Keys
 */
PrismLayer.LayerOptionsKeys = plot_layer_1.PlotLayer.LayerConfigkeys.concat(LAYER_OPTIONS_KEYS);
