// @antv/l7plot/dist/lib/plots/path/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Path = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const path_layer_1 = require("../../layers/path-layer");
const layer_group_1 = require("../../core/layer/layer-group");
class Path extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 图表类型
         */
        this.type = plot_1.Plot.PlotType.Path;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Path.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.pathLayer = new path_layer_1.PathLayer(Object.assign({ source }, (0, util_1.pick)(this.options, path_layer_1.PathLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.pathLayer]);
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const dotLayerConfig = (0, util_1.pick)(options, path_layer_1.PathLayer.LayerOptionsKeys);
        this.pathLayer.update(dotLayerConfig);
    }
    /**
     * 实现 legend 配置项
     */
    getLegendOptions() {
        const colorLegendItems = this.pathLayer.getColorLegendItems();
        if (colorLegendItems.length !== 0) {
            return { type: 'category', items: colorLegendItems };
        }
        return {};
    }
}
exports.Path = Path;
/**
 * 默认配置项
 */
Path.DefaultOptions = constants_1.DEFAULT_OPTIONS;
