// @antv/l7plot/dist/lib/plots/area/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Area = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const area_layer_1 = require("../../layers/area-layer");
const layer_group_1 = require("../../core/layer/layer-group");
class Area extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 图表类型
         */
        this.type = plot_1.Plot.PlotType.Area;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Area.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.areaLayer = new area_layer_1.AreaLayer(Object.assign({ source }, (0, util_1.pick)(this.options, area_layer_1.AreaLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.areaLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const polygonLayerConfig = (0, util_1.pick)(options, area_layer_1.AreaLayer.LayerOptionsKeys);
        this.areaLayer.update(polygonLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
    /**
     * 初始化图层事件
     */
    initLayersEvent() {
        //
    }
    /**
     * 实现 legend 配置项
     */
    getLegendOptions() {
        const colorLegendItems = this.areaLayer.getColorLegendItems();
        if (colorLegendItems.length !== 0) {
            return { type: 'category', items: colorLegendItems };
        }
        return {};
    }
}
exports.Area = Area;
/**
 * 默认配置项
 */
Area.DefaultOptions = constants_1.DEFAULT_OPTIONS;
