// @antv/l7plot/dist/lib/layers/heatmap-layer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeatmapLayer = void 0;
const util_1 = require("@antv/util");
const l7_1 = require("@antv/l7");
const plot_layer_1 = require("../../core/layer/plot-layer");
const adaptor_1 = require("./adaptor");
const DEFAULT_OPTIONS = {
    source: {
        data: [],
        parser: {
            type: 'json',
            x: 'x',
            y: 'y',
        },
    },
    shape: 'heatmap',
    size: {
        value: [0, 1],
    },
    style: {
        intensity: 3,
        radius: 20,
        opacity: 1,
        colorsRamp: [
            { color: 'rgba(33,102,172,0.0)', position: 0 },
            { color: 'rgb(103,169,207)', position: 0.2 },
            { color: 'rgb(209,229,240)', position: 0.4 },
            { color: 'rgb(253,219,199)', position: 0.6 },
            { color: 'rgb(239,138,98)', position: 0.8 },
            { color: 'rgb(178,24,43,1.0)', position: 1 },
        ],
    },
};
const LAYER_OPTIONS_KEYS = ['shape', 'color', 'size', 'style', 'state'];
class HeatmapLayer extends plot_layer_1.PlotLayer {
    constructor(options) {
        super(options);
        /**
         * 图层类型
         */
        this.type = plot_layer_1.PlotLayer.LayerType.HeatmapLayer;
        /**
         * 图层是否具有交互属性
         */
        this.interaction = false;
        const { name, source } = this.options;
        const config = this.pickLayerConfig(this.options);
        this.name = name ? name : (0, util_1.uniqueId)(this.type);
        this.layer = new l7_1.HeatmapLayer(Object.assign(Object.assign({}, config), { name: this.name }));
        this.mappingLayer(this.layer, this.options);
        this.setSource(source);
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return DEFAULT_OPTIONS;
    }
    mappingLayer(layer, options) {
        (0, adaptor_1.mappingLayer)(layer, options);
    }
    update(options) {
        super.update(options);
        this.mappingLayer(this.layer, this.options);
    }
}
exports.HeatmapLayer = HeatmapLayer;
/**
 * 默认配置项
 */
HeatmapLayer.DefaultOptions = DEFAULT_OPTIONS;
/**
 * 图层配置项 Keys
 */
HeatmapLayer.LayerOptionsKeys = plot_layer_1.PlotLayer.LayerConfigkeys.concat(LAYER_OPTIONS_KEYS);
