// @antv/l7plot/dist/esm/plots/hexbin/index.js
import { pick } from '@antv/util';
import { Plot } from '../../core/plot';
import { DEFAULT_OPTIONS } from './constants';
import { HexbinLayer } from '../../layers/hexbin-layer';
import { LayerGroup } from '../../core/layer/layer-group';
export class Hexbin extends Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = Plot.PlotType.Hexbin;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Hexbin.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.hexbinLayer = new HexbinLayer(Object.assign({ source }, pick(this.options, HexbinLayer.LayerOptionsKeys)));
        const layerGroup = new LayerGroup([this.hexbinLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(this.source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const heatMapLayerConfig = pick(options, HexbinLayer.LayerOptionsKeys);
        this.hexbinLayer.update(heatMapLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
}
/**
 * 默认配置项
 */
Hexbin.DefaultOptions = DEFAULT_OPTIONS;
