// @antv/l7plot/dist/lib/plots/dot/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dot = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const dot_layer_1 = require("../../layers/dot-layer");
const layer_group_1 = require("../../core/layer/layer-group");
class Dot extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 图表类型
         */
        this.type = plot_1.Plot.PlotType.Dot;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Dot.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.dotLayer = new dot_layer_1.DotLayer(Object.assign({ source }, (0, util_1.pick)(this.options, dot_layer_1.DotLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.dotLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const dotLayerConfig = (0, util_1.pick)(options, dot_layer_1.DotLayer.LayerOptionsKeys);
        this.dotLayer.update(dotLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
    /**
     * 实现 legend 配置项
     */
    getLegendOptions() {
        const colorLegendItems = this.dotLayer.getColorLegendItems();
        if (colorLegendItems.length !== 0) {
            return { type: 'category', items: colorLegendItems };
        }
        return {};
    }
}
exports.Dot = Dot;
/**
 * 默认配置项
 */
Dot.DefaultOptions = constants_1.DEFAULT_OPTIONS;
