// @antv/l7plot/dist/esm/component/legend.js
import { Control } from '@antv/l7';
import { CategoryLegend, ContinueLegend } from '@antv/l7plot-component';
export class Legend extends Control {
    constructor(options) {
        super(options);
        /**
         * legendComponents 实例
         */
        this.legendComponents = [];
        this.options = options;
        this.legendComponents = this.initLegendComponents(options.items);
    }
    initLegendComponents(legendItem) {
        const legendComponents = [];
        for (let index = 0; index < legendItem.length; index++) {
            const item = legendItem[index];
            if (item.type === 'category') {
                const legend = new CategoryLegend({
                    title: item.options.title,
                    items: item.options.items,
                    className: item.options.className,
                    customContent: item.options.customContent,
                    domStyles: item.options.domStyles,
                });
                legendComponents.push(legend);
            }
            else if (item.type === 'continue') {
                const legend = new ContinueLegend({
                    title: item.options.title,
                    min: item.options.min,
                    max: item.options.max,
                    colors: item.options.colors,
                    className: item.options.className,
                    customContent: item.options.customContent,
                    domStyles: item.options.domStyles,
                });
                legendComponents.push(legend);
            }
        }
        return legendComponents;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return {
            position: 'bottomleft',
        };
    }
    onAdd() {
        const container = window.document.createElement('div');
        container.className = 'l7plot-legend-container';
        this.legendComponents.forEach((legendComponent) => {
            const legend = legendComponent.getContainer();
            container.appendChild(legend);
        });
        return container;
    }
    onRemove() {
        this.legendComponents.forEach((legendComponent) => {
            legendComponent.destroy();
        });
    }
}
