// @antv/l7plot/dist/esm/plots/grid/index.js
import { pick } from '@antv/util';
import { Plot } from '../../core/plot';
import { DEFAULT_OPTIONS } from './constants';
import { GridLayer } from '../../layers/grid-layer';
import { LayerGroup } from '../../core/layer/layer-group';
export class Grid extends Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = Plot.PlotType.Grid;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Grid.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.gridLayer = new GridLayer(Object.assign({ source }, pick(this.options, GridLayer.LayerOptionsKeys)));
        const layerGroup = new LayerGroup([this.gridLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(this.source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const heatMapLayerConfig = pick(options, GridLayer.LayerOptionsKeys);
        this.gridLayer.update(heatMapLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
}
/**
 * 默认配置项
 */
Grid.DefaultOptions = DEFAULT_OPTIONS;
