// @antv/l7plot/dist/lib/plots/hexbin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hexbin = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const hexbin_layer_1 = require("../../layers/hexbin-layer");
const layer_group_1 = require("../../core/layer/layer-group");
class Hexbin extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = plot_1.Plot.PlotType.Hexbin;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return Hexbin.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.hexbinLayer = new hexbin_layer_1.HexbinLayer(Object.assign({ source }, (0, util_1.pick)(this.options, hexbin_layer_1.HexbinLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.hexbinLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(this.source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const heatMapLayerConfig = (0, util_1.pick)(options, hexbin_layer_1.HexbinLayer.LayerOptionsKeys);
        this.hexbinLayer.update(heatMapLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    }
}
exports.Hexbin = Hexbin;
/**
 * 默认配置项
 */
Hexbin.DefaultOptions = constants_1.DEFAULT_OPTIONS;
