// @antv/l7plot-component/dist/lib/legend/continue/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const theme_1 = tslib_1.__importDefault(require("../../configs/theme"));
const constants_1 = require("./constants");
exports.default = {
    // css style for legend
    [constants_1.CONTAINER_CLASS]: {
        visibility: 'visible',
        zIndex: 1,
        backgroundColor: 'rgba(255, 255, 255, 0.9)',
        boxShadow: 'rgb(0 0 0 / 16%) 0px 6px 12px 0px',
        borderRadius: '2px',
        color: 'rgb(87, 87, 87)',
        fontFamily: theme_1.default.fontFamily,
        padding: '10px',
        lineHeight: 1,
        fontSize: '12px',
    },
    [constants_1.TITLE_CLASS]: {
        fontSize: '13px',
        lineHeight: '19px',
        marginBottom: '8px',
    },
    [constants_1.RIBBON_CLASS]: {
        display: 'flex',
        alignItems: 'center',
    },
    [constants_1.GRADIENT_BAR_CLASS]: {
        width: '140px',
        height: '14px',
        margin: '0px 5px',
    },
    [constants_1.VALUE_RANGE_CLASS]: {
        padding: '0px',
    },
};
