// @antv/l7plot/dist/esm/plots/dot-density/index.js
import { pick } from '@antv/util';
import { Plot } from '../../core/plot';
import { DEFAULT_OPTIONS } from './constants';
import { LayerGroup } from '../../core/layer/layer-group';
import { DotDensityLayer } from '../../layers/dot-density-layer';
export class DotDensity extends Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = Plot.PlotType.DotDensity;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return DotDensity.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.dotDensityLayer = new DotDensityLayer(Object.assign({ source }, pick(this.options, DotDensityLayer.LayerOptionsKeys)));
        const layerGroup = new LayerGroup([this.dotDensityLayer]);
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const bubbleLayerConfig = pick(options, DotDensityLayer.LayerOptionsKeys);
        this.dotDensityLayer.update(bubbleLayerConfig);
    }
}
/**
 * 默认配置项
 */
DotDensity.DefaultOptions = DEFAULT_OPTIONS;
