// @antv/l7plot/dist/lib/plots/dot-density/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DotDensity = void 0;
const util_1 = require("@antv/util");
const plot_1 = require("../../core/plot");
const constants_1 = require("./constants");
const layer_group_1 = require("../../core/layer/layer-group");
const dot_density_layer_1 = require("../../layers/dot-density-layer");
class DotDensity extends plot_1.Plot {
    constructor() {
        super(...arguments);
        /**
         * 地图类型
         */
        this.type = plot_1.Plot.PlotType.DotDensity;
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return DotDensity.DefaultOptions;
    }
    /**
     * 创建图层
     */
    createLayers(source) {
        this.dotDensityLayer = new dot_density_layer_1.DotDensityLayer(Object.assign({ source }, (0, util_1.pick)(this.options, dot_density_layer_1.DotDensityLayer.LayerOptionsKeys)));
        const layerGroup = new layer_group_1.LayerGroup([this.dotDensityLayer]);
        return layerGroup;
    }
    /**
     * 更新图层
     */
    updateLayers(options) {
        const bubbleLayerConfig = (0, util_1.pick)(options, dot_density_layer_1.DotDensityLayer.LayerOptionsKeys);
        this.dotDensityLayer.update(bubbleLayerConfig);
    }
}
exports.DotDensity = DotDensity;
/**
 * 默认配置项
 */
DotDensity.DefaultOptions = constants_1.DEFAULT_OPTIONS;
