// @antv/l7plot/dist/esm/layers/text-layer/index.js
import { uniqueId } from '@antv/util';
import { PointLayer } from '@antv/l7';
import { PlotLayer } from '../../core/layer/plot-layer';
import { mappingLayer } from './adaptor';
const DEFAULT_OPTIONS = {
    style: {
        fontSize: 12,
    },
};
const LAYER_OPTIONS_KEYS = [];
export class TextLayer extends PlotLayer {
    constructor(options) {
        super(options);
        /**
         * 图层类型
         */
        this.type = PlotLayer.LayerType.TextLayer;
        /**
         * 图层是否具有交互属性
         */
        this.interaction = false;
        const { name, source } = this.options;
        const config = this.pickLayerConfig(this.options);
        this.name = name ? name : uniqueId(this.type);
        this.layer = new PointLayer(Object.assign(Object.assign({}, config), { name: this.name }));
        this.mappingLayer(this.layer, this.options);
        this.setSource(source);
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return DEFAULT_OPTIONS;
    }
    mappingLayer(layer, options) {
        mappingLayer(layer, options);
    }
    update(options) {
        super.update(options);
        this.mappingLayer(this.layer, this.options);
    }
}
/**
 * 默认配置项
 */
TextLayer.DefaultOptions = DEFAULT_OPTIONS;
/**
 * 图层配置项 Keys
 */
TextLayer.LayerOptionsKeys = PlotLayer.LayerConfigkeys.concat(LAYER_OPTIONS_KEYS);
