// @antv/l7plot/dist/esm/layers/dot-layer/index.js
import { uniqueId } from '@antv/util';
import { PointLayer } from '@antv/l7';
import { PlotLayer } from '../../core/layer/plot-layer';
import { mappingLayer } from './adaptor';
import { getColorLegendItems } from './helper';
const DEFAULT_OPTIONS = {
    source: {
        data: [],
        parser: {
            type: 'json',
            x: 'x',
            y: 'y',
        },
    },
    shape: 'circle',
    size: 12,
    color: '#5FD3A6',
};
const LAYER_OPTIONS_KEYS = ['shape', 'color', 'size', 'style', 'state', 'animate'];
export class DotLayer extends PlotLayer {
    constructor(options) {
        super(options);
        /**
         * 图层类型
         */
        this.type = PlotLayer.LayerType.DotLayer;
        /**
         * 图层是否具有交互属性
         */
        this.interaction = true;
        const { name, source } = this.options;
        const config = this.pickLayerConfig(this.options);
        this.name = name ? name : uniqueId(this.type);
        this.layer = new PointLayer(Object.assign(Object.assign({}, config), { name: this.name }));
        this.mappingLayer(this.layer, this.options);
        this.setSource(source);
    }
    /**
     * 获取默认配置
     */
    getDefaultOptions() {
        return DEFAULT_OPTIONS;
    }
    mappingLayer(layer, options) {
        mappingLayer(layer, options);
    }
    update(options) {
        super.update(options);
        this.mappingLayer(this.layer, this.options);
    }
    getColorLegendItems() {
        const colorLegendItems = this.layer.getLegendItems('color');
        if (Array.isArray(colorLegendItems) && colorLegendItems.length !== 0) {
            const items = getColorLegendItems(colorLegendItems);
            return items;
        }
        return [];
    }
}
/**
 * 默认配置项
 */
DotLayer.DefaultOptions = DEFAULT_OPTIONS;
/**
 * 图层配置项 Keys
 */
DotLayer.LayerOptionsKeys = PlotLayer.LayerConfigkeys.concat(LAYER_OPTIONS_KEYS);
